File: src/core/sse-parser.ts
// Shared SSE event parser for streaming LLM responses.
// Extracted from llm-client.ts — used by AnthropicCompatible and OpenAICompatible clients.
// Pure function (no side effects, no LLM client dependency) — fully unit-testable.

export type SSEFormat = 'anthropic' | 'openai';

interface AnthropicDelta {
  type?: string;
  delta?: { type?: string; text?: string };
}

interface OpenAIDelta {
  choices?: Array<{
    delta?: { content?: string };
    finish_reason?: string;
  }>;
}

export interface SSEDelta {
  /** Text chunk extracted from this event. Empty string if no text content. */
  text: string;
  /** Provider-specific finish signal (e.g. 'stop', 'length'). */
  finishReason?: string;
}

/**
 * Parse a Server-Sent Events response body into per-event text deltas.
 *
 * Handles both `data: ` (with space, official Anthropic) and `data:` (no space,
 * some third-party proxies). Also handles `\r\n` and `\n` line endings.
 *
 * Malformed JSON in individual events is skipped silently (best-effort parsing).
 *
 * @param responseText - Raw SSE response body
 * @param format - Which provider's event format to parse
 * @returns Array of extracted deltas in order
 */
export function parseSSEEvents(responseText: string, format: SSEFormat): SSEDelta[] {
  const normalized = responseText.replace(/\r\n/g, '\n');
  const events = normalized.split('\n\n');
  const deltas: SSEDelta[] = [];

  for (const event of events) {
    if (!event.trim()) continue;

    const dataLine = event.split('\n').find(line => line.startsWith('data:'));
    if (!dataLine) continue;

    if (format === 'openai') {
      const dataContent = dataLine.substring(5).trim();
      // [DONE] is the OpenAI stream terminator — stop processing all subsequent events.
      if (dataContent === '[DONE]') break;
    }

    try {
      const jsonStart = dataLine.indexOf('{');
      if (jsonStart === -1) continue;
      const parsed = JSON.parse(dataLine.substring(jsonStart)) as AnthropicDelta | OpenAIDelta;

      if (format === 'anthropic') {
        const anth = parsed as AnthropicDelta;
        if (anth.type === 'content_block_delta' &&
            anth.delta?.type === 'text_delta' &&
            anth.delta.text) {
          deltas.push({ text: anth.delta.text });
        }
      } else {
        const oai = parsed as OpenAIDelta;
        const content = oai.choices?.[0]?.delta?.content;
        const finish = oai.choices?.[0]?.finish_reason;
        if (content) deltas.push({ text: content });
        if (finish) deltas.push({ text: '', finishReason: finish });
      }
    } catch {
      // Skip malformed JSON in SSE
    }
  }

  return deltas;
}

File: python_lib/llm_wiki/core/sse_parser.py
"""
SSE event parser — port of src/core/sse-parser.ts
Pure function, no IO.
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import List, Literal, Optional


SSEFormat = Literal["anthropic", "openai"]


@dataclass
class SSEDelta:
    text: str = ""
    reasoning: Optional[str] = None
    finish_reason: Optional[str] = None


def parse_sse_events(response_text: str, format: SSEFormat) -> List[SSEDelta]:
    """Parse a Server-Sent Events response body into per-event text deltas."""
    normalized = response_text.replace("\r\n", "\n")
    events = normalized.split("\n\n")
    deltas: List[SSEDelta] = []

    for event in events:
        if not event.strip():
            continue

        lines = event.split("\n")
        data_line = next((l for l in lines if l.startswith("data:")), None)
        if not data_line:
            continue

        if format == "openai":
            data_content = data_line[5:].strip()
            if data_content == "[DONE]":
                break

        try:
            json_start = data_line.index("{")
            parsed = json.loads(data_line[json_start:])

            if format == "anthropic":
                if (parsed.get("type") == "content_block_delta"
                        and parsed.get("delta", {}).get("type") == "text_delta"):
                    text = parsed["delta"].get("text", "")
                    if text:
                        deltas.append(SSEDelta(text=text))
            else:  # openai
                choices = parsed.get("choices", [])
                if choices:
                    delta = choices[0].get("delta", {})
                    content = delta.get("content")
                    reasoning = delta.get("reasoning_content")
                    finish = choices[0].get("finish_reason")
                    if content or reasoning:
                        deltas.append(SSEDelta(
                            text=content or "",
                            reasoning=reasoning or None,
                        ))
                    if finish:
                        deltas.append(SSEDelta(text="", finish_reason=finish))
        except (ValueError, KeyError):
            # Skip malformed JSON in SSE
            pass

    return deltas
